### PR TITLE
Check if filtration value is finite before recursively prune above filtration value

### LINKED ISSUE #926: Avoid `prune_above_filtration(+inf)`
https://github.com/GUDHI/gudhi-devel/blob/e277777cdf2f6163a94767b2d9a74e46f534cdfa/src/Alpha_complex/include/gudhi/Alpha_complex.h#L479
By default, this calls the useless `prune_above_filtration(+inf)`. I think we should either avoid calling it when the parameter is +inf, or modify `prune_above_filtration` so it returns early in that case.

### PR DESCRIPTION
**Pull request by VincentRouvreau**
Fix #926
~~TBD, but I used [std::isfinite](https://en.cppreference.com/w/cpp/numeric/math/isfinite).~~ (not a good idea, I didn't think about -Inf case)
~~Maybe not conform with [Filtration value concept](https://github.com/GUDHI/gudhi-devel/blob/master/src/Simplex_tree/concept/FiltrationValue.h)~~ (cf. comments)
Check if it is equal to `std::numeric_limits<Filtration_value>::infinity()`.
~~NaN case management (and UT)~~

### CODE REVIEW

File: src/Simplex_tree/include/gudhi/Simplex_tree.h [hunk 1]
```diff
@@ -1575,6 +1576,8 @@ class Simplex_tree {
    * bound. If you care, you can call `dimension()` to recompute the exact dimension.
    */
   bool prune_above_filtration(Filtration_value filtration) {
+    if (!std::isfinite(filtration))
```
Comment (mglisse) [inline src/Simplex_tree/include/gudhi/Simplex_tree.h:1579]: `prune_above_filtration(-INFINITY)` should remove (almost?) everything, I don't think `isfinite` is right.

**Comment by mglisse** [comment]
Ideally we wouldn't need to do anything, the compiler would notice that f<=+inf is always true, remove the test, notice that the code does a read-only walk on part of a graph, and eliminate it. But first, `f<=+inf` is false for NaN. I guess it is good that the function removes simplices with a NaN filtration value (?), so changing it to `!(f>+inf)` to make it always true would be a bad idea.
As usual when using `numeric_limits<X>::infinity`, it may be safer to check the corresponding `has_infinity` first.
What exactly did you have in mind for the concept?

**Comment by VincentRouvreau** [comment]
> What exactly did you have in mind for the concept?

Forget about this point, your advice (*may be safer to check the corresponding `has_infinity`*) fixed it.

**Review by mglisse — COMMENTED** [review]
> As the method is called prune_above_filtration (because of above in the name of the method), it makes non sense to me to do something when calling prune_above_filtration(NaN).

I wasn't thinking of calling the function with NaN, but of what happens if in the Simplex_tree one of the simplices has NaN as filtration value. That's a case we looked at in #172, but we probably don't need to care about that here, let's ignore that possibility.
I wouldn't take the name with "above" too literally, we weren't going to call the function prune_not_below_or_equal_filtration.

File: src/Simplex_tree/include/gudhi/Simplex_tree.h [hunk 2]
```diff
@@ -52,6 +52,7 @@
 #include <iterator>  // for std::distance
 #include <type_traits>  // for std::conditional
 #include <unordered_map>
+#include <cmath>  // for std::isfinite
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/Simplex_tree/include/gudhi/Simplex_tree.h]: Was standing for `std::isnan`, but removed on [b859eac](https://github.com/GUDHI/gudhi-devel/pull/928/commits/b859eac9f9aebc06d48b03ae4a3cfce4dfcd8f60)

File: src/Simplex_tree/include/gudhi/Simplex_tree.h [hunk 3]
```diff
@@ -1575,6 +1576,10 @@ class Simplex_tree {
    * bound. If you care, you can call `dimension()` to recompute the exact dimension.
    */
   bool prune_above_filtration(Filtration_value filtration) {
+    if (std::numeric_limits<Filtration_value>::has_infinity && filtration == std::numeric_limits<Filtration_value>::infinity())
+      return false;  // ---->>
+    if (std::numeric_limits<Filtration_value>::has_quiet_NaN && std::isnan(filtration))
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/Simplex_tree/include/gudhi/Simplex_tree.h]: Ok, I removed NaN management and its UT.
I also added some comments in the UT for `prune_above_filtration(+Inf)`, as this test may become obsolete when we will manage NaN filtration values.
All done on [b859eac](https://github.com/GUDHI/gudhi-devel/pull/928/commits/b859eac9f9aebc06d48b03ae4a3cfce4dfcd8f60)

**Review by mglisse — APPROVED** [review]

**Pull request merged**